{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4a8f95bd-7224-4a1c-b350-5fab83141e21",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "9b063814-eadf-4a17-9c83-ed2418defac8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>instruction</th>\n",
       "      <th>response</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Hi, I have a very basic notebook running in ou...</td>\n",
       "      <td>Cross-referenced post: galaxyproject/admins - ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Any ideas yet what is going on with put and ge...</td>\n",
       "      <td>\\n\\n\\n maikenp:\\n\\njupyter\\n\\n\\nHi @maikenp\\nS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>The key point being I think the use of the qua...</td>\n",
       "      <td>Ok! Progress  Did you set galaxy_infrastructur...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>@hxr do you have some advice on what would be ...</td>\n",
       "      <td>Hi @maikenp sorry for the delay, too many thin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Hi, I met an error when I used the Trinity too...</td>\n",
       "      <td>Hi @luweidong\\nTry transcript assembly for a s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>Hi @wm75\\nI did exactly what you told but i am...</td>\n",
       "      <td>I’m sorry, but I don’t understand what the iss...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>I have a problem when trying to get genomic DN...</td>\n",
       "      <td>Hi @Mattiaf!\\nI noticed a problem myself with ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>Thanks Jennifer.\\nI get the “Chromosome by nam...</td>\n",
       "      <td>Thanks for info. Am reviewing with these extra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>Hello there!\\nI am working with WGBS data and ...</td>\n",
       "      <td>Hi @srinivasa_abishek,\\nI’m trying to reproduc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>The link to the git with the code\\nrecentrifuge</td>\n",
       "      <td>Hi @pimarin,\\nin the from_work_dir option you ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>356 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           instruction  \\\n",
       "0    Hi, I have a very basic notebook running in ou...   \n",
       "1    Any ideas yet what is going on with put and ge...   \n",
       "2    The key point being I think the use of the qua...   \n",
       "3    @hxr do you have some advice on what would be ...   \n",
       "4    Hi, I met an error when I used the Trinity too...   \n",
       "..                                                 ...   \n",
       "351  Hi @wm75\\nI did exactly what you told but i am...   \n",
       "352  I have a problem when trying to get genomic DN...   \n",
       "353  Thanks Jennifer.\\nI get the “Chromosome by nam...   \n",
       "354  Hello there!\\nI am working with WGBS data and ...   \n",
       "355    The link to the git with the code\\nrecentrifuge   \n",
       "\n",
       "                                              response  \n",
       "0    Cross-referenced post: galaxyproject/admins - ...  \n",
       "1    \\n\\n\\n maikenp:\\n\\njupyter\\n\\n\\nHi @maikenp\\nS...  \n",
       "2    Ok! Progress  Did you set galaxy_infrastructur...  \n",
       "3    Hi @maikenp sorry for the delay, too many thin...  \n",
       "4    Hi @luweidong\\nTry transcript assembly for a s...  \n",
       "..                                                 ...  \n",
       "351  I’m sorry, but I don’t understand what the iss...  \n",
       "352  Hi @Mattiaf!\\nI noticed a problem myself with ...  \n",
       "353  Thanks for info. Am reviewing with these extra...  \n",
       "354  Hi @srinivasa_abishek,\\nI’m trying to reproduc...  \n",
       "355  Hi @pimarin,\\nin the from_work_dir option you ...  \n",
       "\n",
       "[356 rows x 2 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(\"../out/data.json\") as fout:\n",
    "    raw_data = json.load(fout)\n",
    "\n",
    "question = []\n",
    "response = []\n",
    "for idx_thread, thread in enumerate(raw_data):\n",
    "    for idx_post in range(int(len(thread) / 2)):\n",
    "        next_posts = thread[idx_post:idx_post+2]\n",
    "        if next_posts[0][\"role\"] == \"user\" and next_posts[1][\"role\"] == \"system\":\n",
    "            question.append(next_posts[0][\"text\"])\n",
    "            response.append(next_posts[1][\"text\"])\n",
    "    if idx_thread == 200:\n",
    "        break\n",
    "\n",
    "# create dataframe\n",
    "q_a_dataframe = pd.DataFrame(zip(question, response), columns=[\"instruction\", \"response\"])\n",
    "q_a_dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "6325eeaa-1fd9-443c-829b-1bc2427fd8d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>instruction</th>\n",
       "      <th>response</th>\n",
       "      <th>prompt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Hi, I have a very basic notebook running in ou...</td>\n",
       "      <td>Cross-referenced post: galaxyproject/admins - ...</td>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Any ideas yet what is going on with put and ge...</td>\n",
       "      <td>\\n\\n\\n maikenp:\\n\\njupyter\\n\\n\\nHi @maikenp\\nS...</td>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>The key point being I think the use of the qua...</td>\n",
       "      <td>Ok! Progress  Did you set galaxy_infrastructur...</td>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>@hxr do you have some advice on what would be ...</td>\n",
       "      <td>Hi @maikenp sorry for the delay, too many thin...</td>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Hi, I met an error when I used the Trinity too...</td>\n",
       "      <td>Hi @luweidong\\nTry transcript assembly for a s...</td>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>Hi @wm75\\nI did exactly what you told but i am...</td>\n",
       "      <td>I’m sorry, but I don’t understand what the iss...</td>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>I have a problem when trying to get genomic DN...</td>\n",
       "      <td>Hi @Mattiaf!\\nI noticed a problem myself with ...</td>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>Thanks Jennifer.\\nI get the “Chromosome by nam...</td>\n",
       "      <td>Thanks for info. Am reviewing with these extra...</td>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>Hello there!\\nI am working with WGBS data and ...</td>\n",
       "      <td>Hi @srinivasa_abishek,\\nI’m trying to reproduc...</td>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>The link to the git with the code\\nrecentrifuge</td>\n",
       "      <td>Hi @pimarin,\\nin the from_work_dir option you ...</td>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>356 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           instruction  \\\n",
       "0    Hi, I have a very basic notebook running in ou...   \n",
       "1    Any ideas yet what is going on with put and ge...   \n",
       "2    The key point being I think the use of the qua...   \n",
       "3    @hxr do you have some advice on what would be ...   \n",
       "4    Hi, I met an error when I used the Trinity too...   \n",
       "..                                                 ...   \n",
       "351  Hi @wm75\\nI did exactly what you told but i am...   \n",
       "352  I have a problem when trying to get genomic DN...   \n",
       "353  Thanks Jennifer.\\nI get the “Chromosome by nam...   \n",
       "354  Hello there!\\nI am working with WGBS data and ...   \n",
       "355    The link to the git with the code\\nrecentrifuge   \n",
       "\n",
       "                                              response  \\\n",
       "0    Cross-referenced post: galaxyproject/admins - ...   \n",
       "1    \\n\\n\\n maikenp:\\n\\njupyter\\n\\n\\nHi @maikenp\\nS...   \n",
       "2    Ok! Progress  Did you set galaxy_infrastructur...   \n",
       "3    Hi @maikenp sorry for the delay, too many thin...   \n",
       "4    Hi @luweidong\\nTry transcript assembly for a s...   \n",
       "..                                                 ...   \n",
       "351  I’m sorry, but I don’t understand what the iss...   \n",
       "352  Hi @Mattiaf!\\nI noticed a problem myself with ...   \n",
       "353  Thanks for info. Am reviewing with these extra...   \n",
       "354  Hi @srinivasa_abishek,\\nI’m trying to reproduc...   \n",
       "355  Hi @pimarin,\\nin the from_work_dir option you ...   \n",
       "\n",
       "                                                prompt  \n",
       "0    Below is an instruction that describes a task....  \n",
       "1    Below is an instruction that describes a task....  \n",
       "2    Below is an instruction that describes a task....  \n",
       "3    Below is an instruction that describes a task....  \n",
       "4    Below is an instruction that describes a task....  \n",
       "..                                                 ...  \n",
       "351  Below is an instruction that describes a task....  \n",
       "352  Below is an instruction that describes a task....  \n",
       "353  Below is an instruction that describes a task....  \n",
       "354  Below is an instruction that describes a task....  \n",
       "355  Below is an instruction that describes a task....  \n",
       "\n",
       "[356 rows x 3 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "template = \"\"\"Below is an instruction that describes a task. Write a response that appropriately completes the request.\n",
    "\n",
    "### Instruction:\n",
    "{}\n",
    "\n",
    "### Response:\\n\"\"\"\n",
    "\n",
    "q_a_dataframe['prompt'] = q_a_dataframe[\"instruction\"].apply(lambda x: template.format(x))\n",
    "q_a_dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "79fb4441-7bce-4a2d-b405-10efc98d2ca6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>prompt</th>\n",
       "      <th>response</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "      <td>Cross-referenced post: galaxyproject/admins - ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "      <td>\\n\\n\\n maikenp:\\n\\njupyter\\n\\n\\nHi @maikenp\\nS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "      <td>Ok! Progress  Did you set galaxy_infrastructur...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "      <td>Hi @maikenp sorry for the delay, too many thin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "      <td>Hi @luweidong\\nTry transcript assembly for a s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "      <td>I’m sorry, but I don’t understand what the iss...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "      <td>Hi @Mattiaf!\\nI noticed a problem myself with ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "      <td>Thanks for info. Am reviewing with these extra...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "      <td>Hi @srinivasa_abishek,\\nI’m trying to reproduc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "      <td>Hi @pimarin,\\nin the from_work_dir option you ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>356 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                prompt  \\\n",
       "0    Below is an instruction that describes a task....   \n",
       "1    Below is an instruction that describes a task....   \n",
       "2    Below is an instruction that describes a task....   \n",
       "3    Below is an instruction that describes a task....   \n",
       "4    Below is an instruction that describes a task....   \n",
       "..                                                 ...   \n",
       "351  Below is an instruction that describes a task....   \n",
       "352  Below is an instruction that describes a task....   \n",
       "353  Below is an instruction that describes a task....   \n",
       "354  Below is an instruction that describes a task....   \n",
       "355  Below is an instruction that describes a task....   \n",
       "\n",
       "                                              response  \n",
       "0    Cross-referenced post: galaxyproject/admins - ...  \n",
       "1    \\n\\n\\n maikenp:\\n\\njupyter\\n\\n\\nHi @maikenp\\nS...  \n",
       "2    Ok! Progress  Did you set galaxy_infrastructur...  \n",
       "3    Hi @maikenp sorry for the delay, too many thin...  \n",
       "4    Hi @luweidong\\nTry transcript assembly for a s...  \n",
       "..                                                 ...  \n",
       "351  I’m sorry, but I don’t understand what the iss...  \n",
       "352  Hi @Mattiaf!\\nI noticed a problem myself with ...  \n",
       "353  Thanks for info. Am reviewing with these extra...  \n",
       "354  Hi @srinivasa_abishek,\\nI’m trying to reproduc...  \n",
       "355  Hi @pimarin,\\nin the from_work_dir option you ...  \n",
       "\n",
       "[356 rows x 2 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q_a_dataframe = q_a_dataframe[['prompt', 'response']]\n",
    "q_a_dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f09c50a8-74b8-41fc-ad6a-e8250653728c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>351</th>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>352</th>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>353</th>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>354</th>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>Below is an instruction that describes a task....</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>356 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  text\n",
       "0    Below is an instruction that describes a task....\n",
       "1    Below is an instruction that describes a task....\n",
       "2    Below is an instruction that describes a task....\n",
       "3    Below is an instruction that describes a task....\n",
       "4    Below is an instruction that describes a task....\n",
       "..                                                 ...\n",
       "351  Below is an instruction that describes a task....\n",
       "352  Below is an instruction that describes a task....\n",
       "353  Below is an instruction that describes a task....\n",
       "354  Below is an instruction that describes a task....\n",
       "355  Below is an instruction that describes a task....\n",
       "\n",
       "[356 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(None, (356, 1))"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q_a_dataframe['text'] = q_a_dataframe[\"prompt\"] + q_a_dataframe[\"response\"]\n",
    "q_a_dataframe.drop(columns=['prompt', 'response'], inplace=True)\n",
    "display(q_a_dataframe), q_a_dataframe.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c691a856-202c-4937-9ca2-5dc6088aec82",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/users/anup/miniconda3/envs/finetune-dnabert2/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "from datasets import load_dataset\n",
    "from datasets import Dataset\n",
    "dataset = Dataset.from_pandas(q_a_dataframe).train_test_split(test_size=0.05, seed=42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "caad485f-1d7c-4519-8b26-356b7ba99b79",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<bound method Module.modules of BioGptForCausalLM(\n",
      "  (biogpt): BioGptModel(\n",
      "    (embed_tokens): Embedding(42384, 1024, padding_idx=1)\n",
      "    (embed_positions): BioGptLearnedPositionalEmbedding(1026, 1024)\n",
      "    (layers): ModuleList(\n",
      "      (0-23): 24 x BioGptDecoderLayer(\n",
      "        (self_attn): BioGptAttention(\n",
      "          (k_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
      "          (v_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
      "          (q_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
      "          (out_proj): Linear(in_features=1024, out_features=1024, bias=True)\n",
      "        )\n",
      "        (activation_fn): GELUActivation()\n",
      "        (self_attn_layer_norm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)\n",
      "        (fc1): Linear(in_features=1024, out_features=4096, bias=True)\n",
      "        (fc2): Linear(in_features=4096, out_features=1024, bias=True)\n",
      "        (final_layer_norm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)\n",
      "      )\n",
      "    )\n",
      "    (layer_norm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)\n",
      "  )\n",
      "  (output_projection): Linear(in_features=1024, out_features=42384, bias=False)\n",
      ")>\n"
     ]
    }
   ],
   "source": [
    "import torch\n",
    "from transformers import BioGptTokenizer, BioGptForCausalLM, set_seed\n",
    "import re\n",
    "import sys\n",
    "\n",
    "model_path = \"microsoft/biogpt\"\n",
    "\n",
    "tokenizer = BioGptTokenizer.from_pretrained(model_path)\n",
    "model = BioGptForCausalLM.from_pretrained(model_path, torch_dtype=torch.float32)\n",
    "\n",
    "model_modules = str(model.modules)\n",
    "print(model_modules)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "1e593754-a585-47c8-9065-be3a1a1ae745",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trainable params: 786,432 || all params: 347,549,696 || trainable%: 0.22627900672944337\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Map: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 338/338 [00:02<00:00, 131.88 examples/s]\n",
      "Map: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 18/18 [00:00<00:00, 154.54 examples/s]\n",
      "/scratch/users/anup/miniconda3/envs/finetune-dnabert2/lib/python3.9/site-packages/transformers/optimization.py:407: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "    <div>\n",
       "      \n",
       "      <progress value='105' max='105' style='width:300px; height:20px; vertical-align: middle;'></progress>\n",
       "      [105/105 07:19, Epoch 4/5]\n",
       "    </div>\n",
       "    <table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       " <tr style=\"text-align: left;\">\n",
       "      <th>Epoch</th>\n",
       "      <th>Training Loss</th>\n",
       "      <th>Validation Loss</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>5.415300</td>\n",
       "      <td>5.041604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>5.409200</td>\n",
       "      <td>5.028563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>5.391500</td>\n",
       "      <td>5.018176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>5.140400</td>\n",
       "      <td>5.011297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>5.305500</td>\n",
       "      <td>5.009208</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table><p>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "TrainOutput(global_step=105, training_loss=5.3308263142903645, metrics={'train_runtime': 441.5929, 'train_samples_per_second': 3.827, 'train_steps_per_second': 0.238, 'total_flos': 644394528595968.0, 'train_loss': 5.3308263142903645, 'epoch': 4.94})"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from peft import get_peft_config, PeftModel, PeftConfig, get_peft_model, LoraConfig, TaskType\n",
    "import torch\n",
    "from transformers.trainer_callback import TrainerCallback\n",
    "import os\n",
    "import sys\n",
    "from transformers import BitsAndBytesConfig\n",
    "from trl import SFTTrainer\n",
    "\n",
    "target_modules = ['q_proj','v_proj']\n",
    "\n",
    "lora_config = LoraConfig(\n",
    "    r=8,\n",
    "    lora_alpha=8,\n",
    "    lora_dropout=0.05,\n",
    "    bias=\"none\",\n",
    "    target_modules = target_modules,\n",
    "    task_type=\"CAUSAL_LM\",\n",
    ")\n",
    "\n",
    "base_dir = \"biogpt\"\n",
    "\n",
    "per_device_train_batch_size = 4\n",
    "gradient_accumulation_steps = 4\n",
    "optim = 'adamw_hf'\n",
    "learning_rate = 1e-5\n",
    "max_grad_norm = 0.3\n",
    "warmup_ratio = 0.03\n",
    "lr_scheduler_type = \"linear\"\n",
    "\n",
    "\n",
    "from transformers import TrainingArguments\n",
    "training_args = TrainingArguments(\n",
    "    output_dir=base_dir,\n",
    "    save_strategy=\"epoch\",\n",
    "    evaluation_strategy=\"epoch\",\n",
    "    num_train_epochs = 5.0,\n",
    "    logging_strategy=\"epoch\",\n",
    "    logging_steps=100,\n",
    "    per_device_train_batch_size=per_device_train_batch_size,\n",
    "    gradient_accumulation_steps=gradient_accumulation_steps,\n",
    "    optim=optim,\n",
    "    learning_rate=learning_rate,\n",
    "    max_grad_norm=max_grad_norm,\n",
    "    warmup_ratio=warmup_ratio,\n",
    "    group_by_length=True,\n",
    "    lr_scheduler_type=lr_scheduler_type,\n",
    ")\n",
    "    \n",
    "tokenizer.add_special_tokens({'pad_token': '[PAD]'})\n",
    "\n",
    "model = get_peft_model(model, lora_config)\n",
    "model.print_trainable_parameters()\n",
    "\n",
    "trainer = SFTTrainer(\n",
    "    model,\n",
    "    train_dataset=dataset['train'],\n",
    "    eval_dataset = dataset['test'],\n",
    "    dataset_text_field=\"text\",\n",
    "    max_seq_length=256,\n",
    "    args=training_args,\n",
    ")\n",
    "\n",
    "trainer.train()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "ed01b3aa-b450-403e-8d78-40681aac5549",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hello!\n",
      "I am running a local instance of Galaxy (build 22.05). I installed the latest version of Deepvariant (1.4.0+galaxy0) which installed without any errors. However, when I try to run Deepvariant on BAM files output from HISAT2, the error “Fatal error: Exit code 127 ()” comes up. Further, it says that the tool generated the following error: “line 9: run_deepvariant: command not found”.\n",
      "When I look at the backend to see what process Galaxy is going through, even after installation of the tool, the following line keeps repeating on the command line interface:\n",
      "uvicorn.access INFO 2022-12-22 13:43:26,629 [pN:main.1,p:100965,tN:MainThread] 127.0.0.1:58158 - “GET /api/tool_shed_repositories?name=deepvariant&owner=iuc HTTP/1.1” 200\n",
      "I don’t understand this error. Could someone please help me out?\n",
      "I am running the same job on Galaxy.eu server and it is running (for a few hours now) but in the local instance in errors out pretty much instantly.\n",
      "Thanks!\n",
      "\n",
      "response: \n",
      "\n",
      "</s>Below is an instruction that describes a task. Write a response that appropriately completes the request. # # # Instruction: Hello! I am running a local instance of Galaxy (build 22.05). I installed the latest version of Deepvariant (1.4.0 + galaxy0) which installed without any errors. However, when I try to run Deepvariant on BAM files output from HISAT2, the error <unk>Fatal error: Exit code 127 () <unk>comes up. Further, it says that the tool generated the following error: <unk>line 9: run _ deepvariant: command not found <unk>. When I look at the backend to see what process Galaxy is going through, even after installation of the tool, the following line keeps repeating on the command line interface: uvicorn.access INFO 2022-12-22 13: 43: 26,629 [pN: main.1, p: 100965, tN: MainThread] 127.0.0.1: 58158 - <unk>GET / api / tool _ shed _ repositories? name = deepvariant & owner = iuc HTTP / 1.1 <unk>200 I don <unk>t understand this error. Could someone please help me out? I am running the same job on Galaxy.eu server and it is running (for a few hours now) but in the local instance in errors out pretty much instantly. Thanks! # # # Response: Hello! I am running Galaxy. </s>\n",
      "--------\n",
      "Dear Sir,\n",
      "Kindly help in this regards I was trying to make a de novo contig using trinity and it is running since from one week.\n",
      "Is it ok??? or did I out something wrong\n",
      "Kindly help\n",
      "\n",
      "response: \n",
      "\n",
      "</s>Below is an instruction that describes a task. Write a response that appropriately completes the request. # # # Instruction: Dear Sir, Kindly help in this regards I was trying to make a de novo contig using trinity and it is running since from one week. Is it ok??? or did I out something wrong Kindly help # # # Response: Dear Sir, Kindly help in this regards I was trying to make a de novo contig using trinity and it is running since from one week I was trying to make a de novo contig using trinity and it is running since from one week I was trying to make a de novo contig using trinity and it is running since from one week I was trying to make a de novo contig using trinity and it is running since from one week I was trying to make a de novo contig using trinity and it is running since from one week I was trying to make a de novo contig using trinity and it is running since from one week I was trying to make a de novo contig using trinity and it is running since\n",
      "--------\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[37], line 46\u001b[0m\n\u001b[1;32m     43\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m input_string[start_index:end_index]\u001b[38;5;241m.\u001b[39mstrip()\n\u001b[1;32m     45\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;241m3\u001b[39m): \n\u001b[0;32m---> 46\u001b[0m   pred \u001b[38;5;241m=\u001b[39m \u001b[43mpredictions\u001b[49m\u001b[43m[\u001b[49m\u001b[43mi\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m     47\u001b[0m   text \u001b[38;5;241m=\u001b[39m test_strings[i]\n\u001b[1;32m     48\u001b[0m   \u001b[38;5;28mprint\u001b[39m(text\u001b[38;5;241m+\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "# test fine-tuned model\n",
    "# Number 1000, 1001 from data.json\n",
    "test_strings = [\"Hello!\\nI am running a local instance of Galaxy (build 22.05). I installed the latest version of Deepvariant (1.4.0+galaxy0) which installed without any errors. However, when I try to run Deepvariant on BAM files output from HISAT2, the error “Fatal error: Exit code 127 ()” comes up. Further, it says that the tool generated the following error: “line 9: run_deepvariant: command not found”.\\nWhen I look at the backend to see what process Galaxy is going through, even after installation of the tool, the following line keeps repeating on the command line interface:\\nuvicorn.access INFO 2022-12-22 13:43:26,629 [pN:main.1,p:100965,tN:MainThread] 127.0.0.1:58158 - “GET /api/tool_shed_repositories?name=deepvariant&owner=iuc HTTP/1.1” 200\\nI don’t understand this error. Could someone please help me out?\\nI am running the same job on Galaxy.eu server and it is running (for a few hours now) but in the local instance in errors out pretty much instantly.\\nThanks!\",\n",
    "\"Dear Sir,\\nKindly help in this regards I was trying to make a de novo contig using trinity and it is running since from one week.\\nIs it ok??? or did I out something wrong\\nKindly help\",]\n",
    "\n",
    "#response_test_strings = [\n",
    "#\"Hello, I can’t give a full answer but I can maybe guide you in the right direction and maybe someone that can give a better answer will reply.\\nGiven the error it looks like deepvariant is not installed (not found). The tool is using a “docker tool dependency”, in other words it needs a container where deepvariant is installed. If you have not checked this yet then I think this is the place to start. Below two links where you may find some more information.\\nhttps://docs.galaxyproject.org/en/master/admin/special_topics/mulled_containers.html\\n  \\n      \\n\\n      training.galaxyproject.org\\n  \\n\\n  \\n    \\n\\nGalaxy Training: Tool Dependencies and Containers\\n\\n  Galaxy is an open-source project. Everyone can contribute...\\n\\n\\n  \\n\\n  \\n    \\n    \\n  \\n\\n  \\n\\n\\nThe requirement can be seen here:\\n  \\n\\n      github.com\\n  \\n\\n  \\n    galaxyproject/tools-iuc/blob/master/tools/deepvariant/macros.xml\\n\\n\\n      <macros>\\n    <token name=\"@TOOL_VERSION@\">1.4.0</token>\\n    <token name=\"@SUFFIX_VERSION@\">0</token>\\n    <xml name=\"edam_ontology\">\\n        <edam_topics>                                                                                  \\n            <edam_topic>topic_0199</edam_topic>\\n        </edam_topics>\\n        <edam_operations>\\n            <edam_operation>operation_3227</edam_operation>\\n        </edam_operations>\\n    </xml>\\n    <xml name=\"requirements\">\\n        <requirements>\\n            <container type=\"docker\">google/deepvariant:@TOOL_VERSION@</container>\\n        </requirements>\\n    </xml>\\n    <xml name=\"citations\">\\n        <citations>\\n            <citation type=\"doi\">10.1038/nbt.4235</citation>\\n        </citations>\\n\\n\\n\\n\\n  This file has been truncated. show original\\n\\n  \\n\\n  \\n    \\n    \\n  \\n\\n  \\n\\n\", \n",
    "# \"Hello @Sachin_Srivastava\\nIf the job is running (yellow/peach dataset), it is usually best to allow it to run. The same is true for queued jobs (grey dataset). This applies to jobs (any tool) executed at a public Galaxy server.\\n20 GB of fastq data – uncompressed – creates a very large assembly job. If it fails later on for exceeding resources (red dataset), you’ll need to do one or more of these:\\n\\nTry a rerun to eliminate cluster issues\\nMore QA/QC on the input reads (always recommended)\\nConsider downsampling the reads (tool: Seqtk)\\nPossibly need to move to your own Galaxy server where more resources can be allocated. The GVL version of Cloudman is one option: https://launch.usegalaxy.org/catalog\\n\\n\\nI added some tags to your post that will find prior Q&A about the above actions. Or, you can search the forum with those keywords (not all posts get tagged).\\nYou didn’t state where you are working. But, if by chance at Galaxy Main https://usegalaxy.org, I can let you know that the cluster that runs Trinity (and Unicycler + RNA-Star) is very busy. Longer queue times are expected. If you delete the current job and rerun, that will only place your job back at the end of the queue again, extending wait time.\\nThanks!\"]\n",
    "\n",
    "\n",
    "predictions = []\n",
    "for test in test_strings:\n",
    "  prompt = \"\"\"Below is an instruction that describes a task. Write a response that appropriately completes the request.\n",
    "\n",
    "  ### Instruction:\n",
    "  {}\n",
    "\n",
    "  ### Response:\"\"\".format(test)\n",
    "    \n",
    "  input_ids = tokenizer(prompt, return_tensors=\"pt\").input_ids.to('cuda')\n",
    "\n",
    "  generation_output = model.generate(\n",
    "      input_ids=input_ids, max_new_tokens=156\n",
    "  )\n",
    "  predictions.append(tokenizer.decode(generation_output[0]))\n",
    "\n",
    "\n",
    "def extract_response_text(input_string):\n",
    "    start_marker = '### Response:'\n",
    "    end_marker = '###'\n",
    "    \n",
    "    start_index = input_string.find(start_marker)\n",
    "    if start_index == -1:\n",
    "        return None\n",
    "    \n",
    "    start_index += len(start_marker)\n",
    "    \n",
    "    end_index = input_string.find(end_marker, start_index)\n",
    "    if end_index == -1:\n",
    "        return input_string[start_index:]\n",
    "    \n",
    "    return input_string[start_index:end_index].strip()\n",
    "\n",
    "for i in range(3): \n",
    "  pred = predictions[i]\n",
    "  text = test_strings[i]\n",
    "  print(text+'\\n')\n",
    "  print(\"response: \\n\")\n",
    "  print(pred)\n",
    "  #print(extract_response_text(pred))\n",
    "  print('--------')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c96a8e7-09fa-424e-a168-f8d939f8d3ea",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
